### PR TITLE
[clang-format] Do not update cursor pos if no includes replacement

### LINKED ISSUE #77450: [clang-format] Incorrect cursor position with IncludeBlocksStyle Regroup and LineEnding CRLF
Hello,
We recently configured clang-format on our projects with Visual Studio.
We use the version shipped with Visual Studio 2022 : clang-format 16.0.5.

We found what we consider to be a bug: the cursor has the wrong position when formatting a file with CRLF line endings and includes already formatted.

Here are the reproduction steps:

`.clang-format`
```yml
LineEnding: CRLF
IncludeBlocks:   Regroup
IncludeCategories:
  - Regex:           '^<ext/.*\.h>'
    Priority:        2
    SortPriority:    0
  - Regex:           '^<.*\.h>'
    Priority:        1
    SortPriority:    0
  - Regex:           '^<.*'
    Priority:        2
    SortPriority:    0
  - Regex:           '.*\.generated\.h.*' # always last
    Priority:        2
    SortPriority:    100
  - Regex:           '.*\/.*\.h'
    Priority:        20
    SortPriority:    40
  - Regex:           '.*\/.*'
    Priority:        20
    SortPriority:    40
  - Regex:           '.*\.h'
    Priority:        30
    SortPriority:    30
  - Regex:           '.*'
    Priority:        30
    SortPriority:    30
```

`Foo.cpp`
```cpp
#include "Foo.h"
#include "Bar.h"
#include "Sub/Bar.h"
#include "Sub/Foo.h"

class Foo {
}
```

When formatting `Foo.cpp`, it correctly formats the include block:

`Foo.cpp`
```cpp
#include "Foo.h"

#include "Bar.h"

#include "Sub/Bar.h"
#include "Sub/Foo.h"

class Foo {
}
```

But when reformatting again, if the cursor is at pos 62 (just before `#include "Sub/Foo.h"`, note that we use CRLF), it will not return any replacements BUT it will return the wrong position for the cursor: 60. The cursor goes one line up, on `#include "Sub/Bar.h"`. The same thing happens when placing the cursor before `class Foo {`.

This seems to be aggravated by Visual Studio, which always sends to clang-format the start of the line as the cursor position, no matter where the cursor is on the line.

After some research, we think we've found the issue.

`clang/lib/Format/Format.cpp:2694`
```cpp
  std::string result;
  for (unsigned Index : Indices) {
    if (!result.empty()) {
      result += "\n";
      if (Style.IncludeStyle.IncludeBlocks ==
              tooling::IncludeStyle::IBS_Regroup &&
          CurrentCategory != Includes[Index].Category)
        result += "\n";
    }
    result += Includes[Index].Text;
    if (Cursor && CursorIndex == Index)
      *Cursor = IncludesBeginOffset + result.size() - CursorToEOLOffset;
    CurrentCategory = Includes[Index].Category;
  }

  if (Cursor && *Cursor >= IncludesEndOffset)
    *Cursor += result.size() - IncludesBlockSize;

  // If the #includes are out of order, we generate a single replacement fixing
  // the entire range of blocks. Otherwise, no replacement is generated.
  if (replaceCRLF(result) == replaceCRLF(std::string(Code.substr(
                                 IncludesBeginOffset, IncludesBlockSize))))
    return;
```

In this code, both cursor updates seem wrong. They are based on `result.size()` which **does not include the second line ending character** `\r`. The cursor position is then offsetted by the number of lines between include blocks.

The solution might be to use the correct new line character here, or to check the line ending length (1 character for LF and CR, 2 characters for CRLF).

I will provide a PR.

Thanks for your time

### PR DESCRIPTION
**Pull request by NorthBlue333**
Fix https://github.com/llvm/llvm-project/issues/77450.

### CODE REVIEW

File: clang/lib/Format/Format.cpp [hunk 1]
```diff
@@ -3131,20 +3132,25 @@ static void sortCppIncludes(const FormatStyle &Style,
...
                                  IncludesBeginOffset, IncludesBlockSize)))) {
     return;
   }
 
+  if (Cursor) {
+    if (UINT_MAX != NewCursor) {
+      *Cursor = NewCursor;
+    } else if (*Cursor >= IncludesEndOffset) {
+      *Cursor += result.size() - IncludesBlockSize;
+    }
```
Comment (HazardyKnusperkeks) [inline clang/lib/Format/Format.cpp:3151]:
```suggestion
    if (NewCursor != UINT_MAX)
      *Cursor = NewCursor;
    else if (*Cursor >= IncludesEndOffset)
      *Cursor += result.size() - IncludesBlockSize;
```

**Comment by NorthBlue333** [comment]
The change is done 👍 
Is the code formatting check broken?

**Comment by owenca** [comment]
> The change is done 👍 Is the code formatting check broken?

See [here](https://github.com/llvm/llvm-project/pull/76059#issuecomment-1865850011).

**Comment by NorthBlue333** [comment]
Currently fixing the tests.

File: clang/lib/Format/Format.cpp [hunk 2]
```diff
@@ -3132,20 +3133,24 @@ static void sortCppIncludes(const FormatStyle &Style,
...
-
   // If the #includes are out of order, we generate a single replacement fixing
   // the entire range of blocks. Otherwise, no replacement is generated.
   if (replaceCRLF(result) == replaceCRLF(std::string(Code.substr(
                                  IncludesBeginOffset, IncludesBlockSize)))) {
     return;
   }
 
+  if (Cursor) {
+    if (UINT_MAX != NewCursor)
```
Comment (HazardyKnusperkeks) [inline clang/lib/Format/Format.cpp:3148]:
```suggestion
    if (NewCursor != UINT_MAX)
```
Still that. :)

**Comment by NorthBlue333** [comment]
Oh, sorry @HazardyKnusperkeks I missed this in your previous comment.

I've updated the PR.

As a side note, I tried adding these two tests but they fail, meaning the cursor is still incorrectly computed with CRLF when replacing lines.

This is due to adding lines between the include groups (with Regroup option), as this new line is added with only the character `\n`. The line break between include lines is right because the include line itself contains a trailing `\r`.
I am not sure how to fix this correctly because I do not want to break things by trimming the include line and I do not think `sortCppIncludes` should manage whitespaces (`WhitespaceManager` seems to be the correct class to handle this?). Maybe the cursor should be updated when replacing these whitespaces?

```c++
TEST_F(SortIncludesTest,
       CalculatesCorrectCursorPositionWhenNewLineReplacementsWithRegroupingAndCRLF) {
  Style.IncludeBlocks	  = Style.IBS_Regroup;
  FmtStyle.LineEnding	  = FormatStyle::LE_CRLF;
  Style.IncludeCategories = {
	  {"^\"a\"", 0, 0, false},
	  {"^\"b\"", 1, 1, false},
	  {".*",	 2, 2, false}
	 };
  std::string Code =
	  "#include \"a\"\r\n"	// Start of line: 0
	  "#include \"b\"\r\n"	// Start of line: 14
	  "#include \"c\"\r\n"	// Start of line: 28
	  "\r\n"				// Start of line: 42
	  "int i;";				// Start of line: 44
  std::string Expected =
	  "#include \"a\"\r\n"	// Start of line: 0
	  "\r\n"				// Start of line: 14
	  "#include \"b\"\r\n"	// Start of line: 16
	  "\r\n"				// Start of line: 30
	  "#include \"c\"\r\n"	// Start of line: 32
	  "\r\n"				// Start of line: 46
	  "int i;";				// Start of line: 48
  EXPECT_EQ(Expected, sort(Code));
  EXPECT_EQ(0u, newCursor(Code, 0));
  EXPECT_EQ(15u, newCursor(Code, 14)); // FIXME: should expect 16, caused by \r
  EXPECT_EQ(30u, newCursor(Code, 28)); // FIXME: should expect 32, caused by \r
  EXPECT_EQ(44u, newCursor(Code, 42)); // FIXME: should expect 46, caused by \r
  EXPECT_EQ(46u, newCursor(Code, 44)); // FIXME: should expect 48, caused by \r
}

TEST_F(SortIncludesTest,
       CalculatesCorrectCursorPositionWhenNoNewLineReplacementsWithRegroupingAndCRLF) {
  Style.IncludeBlocks = Style.IBS_Regroup;
  FmtStyle.LineEnding = FormatStyle::LE_CRLF;
  Style.IncludeCategories = {
      {"^\"a\"", 0, 0, false}, {"^\"b\"", 1, 1, false}, {".*", 2, 2, false}};
  std::string Code = "#include \"a\"\r\n" // Start of line: 0
                     "\r\n"               // Start of line: 14
                     "#include \"c\"\r\n" // Start of line: 16
                     "\r\n"               // Start of line: 30
                     "#include \"b\"\r\n" // Start of line: 32
                     "\r\n"               // Start of line: 46
                     "int i;";            // Start of line: 48
  std::string Expected = "#include \"a\"\r\n" // Start of line: 0
                         "\r\n"               // Start of line: 14
                         "#include \"b\"\r\n" // Start of line: 16
                         "\r\n"               // Start of line: 30
                         "#include \"c\"\r\n" // Start of line: 32
                         "\r\n"               // Start of line: 46
                         "int i;";            // Start of line: 48
  EXPECT_EQ(Expected, sort(Code));
  EXPECT_EQ(0u, newCursor(Code, 0));
  EXPECT_EQ(14u, newCursor(Code, 14));
  EXPECT_EQ(30u, newCursor(Code, 16)); // FIXME: should expect 32, caused by \r
  EXPECT_EQ(30u, newCursor(Code, 30));
  EXPECT_EQ(15u, newCursor(Code, 32)); // FIXME: should expect 15, caused by \r
  EXPECT_EQ(44u, newCursor(Code, 46)); // FIXME: should expect 46, caused by \r
  EXPECT_EQ(46u, newCursor(Code, 48)); // FIXME: should expect 48, caused by \r
}
```

**Comment by HazardyKnusperkeks** [comment]
> Oh, sorry @HazardyKnusperkeks I missed this in your previous comment.
> 
> I've updated the PR.
> 
> As a side note, I tried adding these two tests but they fail, meaning the cursor is still incorrectly computed with CRLF when replacing lines.
> 
> This is due to adding lines between the include groups (with Regroup option), as this new line is added with only the character `\n` (while the line break between include lines is right because the include line itself contains a trailing `\r`). I am not sure how to fix this correctly because I do not want to break things by trimming the include line and I do not think `sortCppIncludes` should manage whitespaces (`WhitespaceManager` seems to be the correct class to handle this?). Maybe the cursor should be updated when replacing these whitespaces?
> 
> ```c++
> TEST_F(
>        SortIncludesTest,
>        CalculatesCorrectCursorPositionWhenNewLineReplacementsWithRegroupingAndCRLF) {
>   Style.IncludeBlocks	  = Style.IBS_Regroup;
>   FmtStyle.LineEnding	  = FormatStyle::LE_CRLF;
>   Style.IncludeCategories = {
>       {"^\"a\"", 0, 0, false}, {"^\"b\"", 1, 1, false}, {".*", 2, 2, false}};
>   std::string Code = "#include \"a\"\r\n" // Start of line: 0
>                      "#include \"b\"\r\n" // Start of line: 14
>                      "#include \"c\"\r\n" // Start of line: 28
>                      "\r\n"               // Start of line: 42
>                      "int i;";            // Start of line: 44
>   std::string Expected = "#include \"a\"\r\n" // Start of line: 0
>                          "\r\n"               // Start of line: 14
>                          "#include \"b\"\r\n" // Start of line: 16
>                          "\r\n"               // Start of line: 30
>                          "#include \"c\"\r\n" // Start of line: 32
>                          "\r\n"               // Start of line: 46
>                          "int i;";            // Start of line: 48
>   EXPECT_EQ(Expected, sort(Code));
>   EXPECT_EQ(0u, newCursor(Code, 0));
>   EXPECT_EQ(15u, newCursor(Code, 14)); // FIXME: should expect 16, caused by \r
>   EXPECT_EQ(30u, newCursor(Code, 28)); // FIXME: should expect 32, caused by \r
>   EXPECT_EQ(44u, newCursor(Code, 42)); // FIXME: should expect 46, caused by \r
>   EXPECT_EQ(46u, newCursor(Code, 44)); // FIXME: should expect 48, caused by \r
> }
> 
> TEST_F(
>        SortIncludesTest,
>        CalculatesCorrectCursorPositionWhenNoNewLineReplacementsWithRegroupingAndCRLF) {
>   Style.IncludeBlocks = Style.IBS_Regroup;
>   FmtStyle.LineEnding = FormatStyle::LE_CRLF;
>   Style.IncludeCategories = {
>       {"^\"a\"", 0, 0, false}, {"^\"b\"", 1, 1, false}, {".*", 2, 2, false}};
>   std::string Code = "#include \"a\"\r\n" // Start of line: 0
>                      "\r\n"               // Start of line: 14
>                      "#include \"c\"\r\n" // Start of line: 16
>                      "\r\n"               // Start of line: 30
>                      "#include \"b\"\r\n" // Start of line: 32
>                      "\r\n"               // Start of line: 46
>                      "int i;";            // Start of line: 48
>   std::string Expected = "#include \"a\"\r\n" // Start of line: 0
>                          "\r\n"               // Start of line: 14
>                          "#include \"b\"\r\n" // Start of line: 16
>                          "\r\n"               // Start of line: 30
>                          "#include \"c\"\r\n" // Start of line: 32
>                          "\r\n"               // Start of line: 46
>                          "int i;";            // Start of line: 48
>   EXPECT_EQ(Expected, sort(Code));
>   EXPECT_EQ(0u, newCursor(Code, 0));
>   EXPECT_EQ(14u, newCursor(Code, 14));
>   EXPECT_EQ(30u, newCursor(Code, 16)); // FIXME: should expect 32, caused by \r
>   EXPECT_EQ(30u, newCursor(Code, 30));
>   EXPECT_EQ(15u, newCursor(Code, 32)); // FIXME: should expect 15, caused by \r
>   EXPECT_EQ(44u, newCursor(Code, 46)); // FIXME: should expect 46, caused by \r
>   EXPECT_EQ(46u, newCursor(Code, 48)); // FIXME: should expect 48, caused by \r
> }
> ```

That's outside my comfortable zone and I'll refer to @owenca or @mydeveloperday.

**Comment by NorthBlue333** [comment]
`clang-check` builds in local.

**Comment by NorthBlue333** [comment]
Hi there,
Any update on this?
Friendly ping @owenca or @mydeveloperday.
I think we could maybe divide the issue in two @HazardyKnusperkeks?

File: clang/lib/Format/Format.cpp [hunk 3]
```diff
@@ -3123,6 +3123,7 @@ static void sortCppIncludes(const FormatStyle &Style,
   }
 
   std::string result;
+  unsigned NewCursor = UINT_MAX;
```
Comment (owenca) [inline clang/lib/Format/Format.cpp:3126]:
```suggestion
  const auto OldCursor = Cursor ? *Cursor : 0;
```

File: clang/lib/Format/Format.cpp [hunk 4]
```diff
@@ -3134,20 +3135,24 @@ static void sortCppIncludes(const FormatStyle &Style,
...
   }
 
-  if (Cursor && *Cursor >= IncludesEndOffset)
-    *Cursor += result.size() - IncludesBlockSize;
-
   // If the #includes are out of order, we generate a single replacement fixing
   // the entire range of blocks. Otherwise, no replacement is generated.
   if (replaceCRLF(result) == replaceCRLF(std::string(Code.substr(
                                  IncludesBeginOffset, IncludesBlockSize)))) {
     return;
```
Comment (owenca) [inline clang/lib/Format/Format.cpp:3146]:
```suggestion
    if (Cursor)
      *Cursor = OldCursor;
    return;
```

**Review by owenca — COMMENTED** [review]
`git diff Format.cpp` output:
```
--- a/clang/lib/Format/Format.cpp
+++ b/clang/lib/Format/Format.cpp
@@ -3116,6 +3116,7 @@ static void sortCppIncludes(const FormatStyle &Style,
     return;
   }
 
+  const auto OldCursor = Cursor ? *Cursor : 0;
   std::string result;
   for (unsigned Index : Indices) {
     if (!result.empty()) {
@@ -3139,6 +3140,8 @@ static void sortCppIncludes(const FormatStyle &Style,
   // the entire range of blocks. Otherwise, no replacement is generated.
   if (replaceCRLF(result) == replaceCRLF(std::string(Code.substr(
                                  IncludesBeginOffset, IncludesBlockSize)))) {
+    if (Cursor)
+      *Cursor = OldCursor;
     return;
   }
 
```

**Comment by owenca** [comment]
> As a side note, I tried adding these two tests but they fail, meaning the cursor is still incorrectly computed with CRLF when replacing lines.
> 
> This is due to adding lines between the include groups (with Regroup option), as this new line is added with only the character `\n` (while the line break between include lines is right because the include line itself contains a trailing `\r`). I am not sure how to fix this correctly because I do not want to break things by trimming the include line and I do not think `sortCppIncludes` should manage whitespaces (`WhitespaceManager` seems to be the correct class to handle this?). Maybe the cursor should be updated when replacing these whitespaces?
> 
> ```c++
> TEST_F(
>        SortIncludesTest,
>        CalculatesCorrectCursorPositionWhenNewLineReplacementsWithRegroupingAndCRLF) {
>   Style.IncludeBlocks	  = Style.IBS_Regroup;
>   FmtStyle.LineEnding	  = FormatStyle::LE_CRLF;
>   Style.IncludeCategories = {
>       {"^\"a\"", 0, 0, false}, {"^\"b\"", 1, 1, false}, {".*", 2, 2, false}};
>   std::string Code = "#include \"a\"\r\n" // Start of line: 0
>                      "#include \"b\"\r\n" // Start of line: 14
>                      "#include \"c\"\r\n" // Start of line: 28
>                      "\r\n"               // Start of line: 42
>                      "int i;";            // Start of line: 44
>   std::string Expected = "#include \"a\"\r\n" // Start of line: 0
>                          "\r\n"               // Start of line: 14
>                          "#include \"b\"\r\n" // Start of line: 16
>                          "\r\n"               // Start of line: 30
>                          "#include \"c\"\r\n" // Start of line: 32
>                          "\r\n"               // Start of line: 46
>                          "int i;";            // Start of line: 48
>   EXPECT_EQ(Expected, sort(Code));
>   EXPECT_EQ(0u, newCursor(Code, 0));
>   EXPECT_EQ(15u, newCursor(Code, 14)); // FIXME: should expect 16, caused by \r
>   EXPECT_EQ(30u, newCursor(Code, 28)); // FIXME: should expect 32, caused by \r
>   EXPECT_EQ(44u, newCursor(Code, 42)); // FIXME: should expect 46, caused by \r
>   EXPECT_EQ(46u, newCursor(Code, 44)); // FIXME: should expect 48, caused by \r
> }
> 
> TEST_F(
>        SortIncludesTest,
>        CalculatesCorrectCursorPositionWhenNoNewLineReplacementsWithRegroupingAndCRLF) {
>   Style.IncludeBlocks = Style.IBS_Regroup;
>   FmtStyle.LineEnding = FormatStyle::LE_CRLF;
>   Style.IncludeCategories = {
>       {"^\"a\"", 0, 0, false}, {"^\"b\"", 1, 1, false}, {".*", 2, 2, false}};
>   std::string Code = "#include \"a\"\r\n" // Start of line: 0
>                      "\r\n"               // Start of line: 14
>                      "#include \"c\"\r\n" // Start of line: 16
>                      "\r\n"               // Start of line: 30
>                      "#include \"b\"\r\n" // Start of line: 32
>                      "\r\n"               // Start of line: 46
>                      "int i;";            // Start of line: 48
>   std::string Expected = "#include \"a\"\r\n" // Start of line: 0
>                          "\r\n"               // Start of line: 14
>                          "#include \"b\"\r\n" // Start of line: 16
>                          "\r\n"               // Start of line: 30
>                          "#include \"c\"\r\n" // Start of line: 32
>                          "\r\n"               // Start of line: 46
>                          "int i;";            // Start of line: 48
>   EXPECT_EQ(Expected, sort(Code));
>   EXPECT_EQ(0u, newCursor(Code, 0));
>   EXPECT_EQ(14u, newCursor(Code, 14));
>   EXPECT_EQ(30u, newCursor(Code, 16)); // FIXME: should expect 32, caused by \r
>   EXPECT_EQ(30u, newCursor(Code, 30));
>   EXPECT_EQ(15u, newCursor(Code, 32)); // FIXME: should expect 15, caused by \r
>   EXPECT_EQ(44u, newCursor(Code, 46)); // FIXME: should expect 46, caused by \r
>   EXPECT_EQ(46u, newCursor(Code, 48)); // FIXME: should expect 48, caused by \r
> }
> ```

With my suggested fix, these tests should also pass.

**Comment by NorthBlue333** [comment]
Woops, sorry, I missed the update here.

I've done all your changes in a second commit just to be able to keep track of my previous changes; but I am actually not sure how your suggested modifications are any different from mine? Well, I admit it looks a bit better, but it works the same? Or am I missing something? You are just storing the old value of Cursor and resetting it when no changes, while I was doing it the other way around: storing the new value and updating it only if no changes.

**Comment by NorthBlue333** [comment]
Friendly ping again here, as the status on this has been asked again by my team! :)
@owenca, @mydeveloperday, and maybe @HazardyKnusperkeks (about splitting the issue in 2).

**Comment by owenca** [comment]
> I've done all your changes in a second commit just to be able to keep track of my previous changes; but I am actually not sure how your suggested modifications are any different from mine? Well, I admit it looks a bit better, but it works the same? Or am I missing something? You are just storing the old value of Cursor and resetting it when no changes, while I was doing it the other way around: storing the new value and updating it only if no changes.

It's been a while, but IIRC your original fix was not equivalent to my suggestion, which IMO is simpler and cleaner.

> It does not fix the tests I mentioned.

My bad. I copy-pasted the two tests you said would fail and ran FormatTests with my suggested fix, and they passed. Of course, I missed the comments in your snippet, e.g.:
```
  EXPECT_EQ(15u, newCursor(Code, 14)); // FIXME: should expect 16, caused by \r
```

**Review by owenca — APPROVED** [review]

**Comment by NorthBlue333** [comment]
Sure, that was was not very clear from my side I agree.

Yes, your implementation is cleaner (I don't know why I did not do that from the beginning, I think I was struggling to fix the other tests and ended up with something more complicated than necessary).

I have squashed the commits in only one.
Thanks for your time!

**Comment by owenca** [comment]
> I have squashed the commits in only one. Note that I have left the failing tests in the commit, I am not sure if I should remove them or not.

Unfortunately, this wiped out my updates that fixed a formatting error and added `#if 0` for the failing tests.

**Review by owenca — APPROVED** [review]

**Pull request merged**